### DISCUSSION TITLE
Get the selected text data.

### DISCUSSION THREAD

**Discussion by SethLK** [espresso3389/pdfrx#265] [Q&A]
When text is selected, how do you get that data without copying it?
How do I implement the custom selection toolbar? For instance, other PDF viewer apps have these:
- copy
- select all
- etc

after selected.

**Comment by espresso3389** [comment]
On the today's Flutter 3.27 release, Flutter introduces a way to get [SelectableRegionState](https://api.flutter.dev/flutter/widgets/SelectableRegionState-class.html) from SelectionArea using GlobalKey.

![image](https://github.com/user-attachments/assets/19f4f4c0-42cc-4b51-8b3e-7a346e0e970b)

[SelectableRegionState](https://api.flutter.dev/flutter/widgets/SelectableRegionState-class.html) has the following methods:

- [clearSelection](https://api.flutter.dev/flutter/widgets/SelectableRegionState/clearSelection.html)
- [selectAll](https://api.flutter.dev/flutter/widgets/SelectableRegionState/selectAll.html)

But for copy, [copySelection](https://api.flutter.dev/flutter/widgets/SelectableRegionState/copySelection.html) is recently deprecated because of the iOS copy notification issue.

And, because pdfrx internally embed [SelectionArea](https://api.flutter.dev/flutter/material/SelectionArea-class.html) inside PdfViewer, you should use [PdfViewerParams.perPageSelectableRegionInjector](https://pub.dev/documentation/pdfrx/1.0.94/pdfrx/PdfViewerParams/perPageSelectableRegionInjector.html) to replace the embedded one like this:

```dart
final selectionAreaKey = GlobalKey<SelectableRegionState>();

PdfViewer.file(
  ...,
  params: PdfViewerParams(
    perPageSelectableRegionInjector: (context, child, page, pageRect) {
      return SelectionArea(
        key: selectionAreaKey,
        onSelectionChanged: (selection) { /* some code to handle selection change */ },
        child: child,
      );
    }
  ),
  ...

// clear selection
final selectableRegionState = selectionAreaKey.state;
state?.clearSelection();
```


Of course, pdfrx has [PdfViewerParams.onTextSelectionChange](https://pub.dev/documentation/pdfrx/1.0.94/pdfrx/PdfViewerParams/onTextSelectionChange.html) method to be notified when text selection change, which has better control on PDF structure.
